crates/runtime/src/dataconnector/github/github_app_token_provider.rs

```diff
@@ -0,0 +1,339 @@
+    UnableToGetSystemTime { source: std::time::SystemTimeError },
+
+    #[snafu(display("Invalid system time"))]
+    InvalidSystemTime { source: std::num::TryFromIntError },
+
+    #[snafu(display("Unable to generate JWT"))]
+    UnableToGenerateJWT { source: jsonwebtoken::errors::Error },
+
+    #[snafu(display("Unable to get GitHub installation access token"))]
+    UnableToGetGitHubInstallationAccessToken { source: reqwest::Error },
+
+    #[snafu(display("Unable to get GitHub installation access token body"))]
+    UnableToGetGitHubInstallationAccessTokenBody { source: reqwest::Error },
+}
+
+#[derive(Serialize)]
+struct Claims {
+    iat: usize,
+    exp: usize,
+    iss: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct TokenResponse {
+    token: String,
+    expires_at: String,
+}
+
+#[async_trait]
+trait TokenGenerator: Send + Sync {
+    async fn generate_token(
+        &self,
+        app_client_id: Arc<str>,
+        private_key: Arc<str>,
+        installation_id: Arc<str>,
+    ) -> Result<TokenResponse, GitHubAppError>;
+}
+
+struct GitHubTokenGenerator {}
+
+#[async_trait]
+impl TokenGenerator for GitHubTokenGenerator {
+    async fn generate_token(
+        &self,
+        app_client_id: Arc<str>,
+        private_key: Arc<str>,
+        installation_id: Arc<str>,
+    ) -> Result<TokenResponse, GitHubAppError> {
+        let iat = usize::try_from(
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .context(UnableToGetSystemTimeSnafu {})?
+                .as_secs(),
+        )
+        .context(InvalidSystemTimeSnafu {})?;
+
+        let exp = iat + 600;
+        let claims = Claims {
+            iat,
+            exp,
+            iss: app_client_id.to_string(),
+        };
+        let private_key = private_key.as_ref();
+        let encoding_key =
+            EncodingKey::from_rsa_pem(private_key.as_bytes()).context(InvalidPrivateKeySnafu {})?;
+
+        let jwt_token = encode(&Header::new(Algorithm::RS256), &claims, &encoding_key)
+            .context(UnableToGenerateJWTSnafu {})?;
+
+        let client = reqwest::Client::new();
+
+        let response = client
+            .post(format!(
+                "https://api.github.com/app/installations/{}/access_tokens",
+                installation_id
+            ))
+            .header("Accept", "application/vnd.github+json")
+            .header("Authorization", format!("Bearer {jwt_token}"))
+            .header("X-GitHub-Api-Version", "2022-11-28")
+            .header("User-Agent", "spice")
+            .send()
+            .await
+            .context(UnableToGetGitHubInstallationAccessTokenSnafu {})?;
+
+        let token_response: TokenResponse = response
+            .json()
+            .await
+            .context(UnableToGetGitHubInstallationAccessTokenBodySnafu {})?;
+
+        // self.expires_at
+        //     .write()
+        //     .await
+        //     .clone_from(&token_response.expires_at);
```

Comment:
Isn't this required?

Reply:
No, that is moved to GitHubAppTokenProvider `get_token` implementation.
`TokenGenerator` encapsulates JWT generation and call to GitHub api.

Reply:
removed the comment